controller_interface/include/controller_interface/async_function_handler.hpp

```diff
@@ -0,0 +1,242 @@
+ */
+template <typename T>
+class AsyncFunctionHandler
+{
+public:
+  AsyncFunctionHandler() = default;
+
+  ~AsyncFunctionHandler() { stop_async_update(); }
+
+  /// Initialize the AsyncFunctionHandler with the get_state_function and async_function
+  /**
+   * @param get_state_function Function that returns the current lifecycle state
+   * @param async_function Function that will be called asynchronously
+   * If the AsyncFunctionHandler is already initialized and is running, it will throw a runtime
+   * error.
+   * If the parsed functions are not valid, it will throw a runtime error.
+   */
+  void init(
+    std::function<const rclcpp_lifecycle::State &()> get_state_function,
+    std::function<T(const rclcpp::Time &, const rclcpp::Duration &)> async_function)
+  {
+    if (get_state_function == nullptr)
+    {
+      throw std::runtime_error(
+        "AsyncFunctionHandler: parsed function to get the lifecycle state is not valid!");
+    }
+    if (async_function == nullptr)
+    {
+      throw std::runtime_error(
+        "AsyncFunctionHandler: parsed function to call asynchronously is not valid!");
+    }
+    if (thread_.joinable())
+    {
+      throw std::runtime_error(
+        "AsyncFunctionHandler: Cannot reinitialize while the thread is "
+        "running. Please stop the async update first!");
+    }
+    get_state_function_ = get_state_function;
+    async_function_ = async_function;
+  }
+
+  /// Triggers the async update method cycle
+  /**
+   * @param time Current time
+   * @param period Current period
+   * @return A pair with the first element being a boolean indicating if the async update method was
+   * triggered and the second element being the last return value of the async function.
+   * If the AsyncFunctionHandler is not initialized properly, it will throw a runtime error.
+   * If the async update method is waiting for the trigger, it will notify the async thread to start
+   * the update cycle.
+   * If the async update method is still running, it will return the last return value of the async
+   * function.
+   *
+   * \note In the case of controllers, The controller manager is responsible
+   * for triggering and maintaining the controller's update rate, as it should be only acting as a
+   * scheduler. Same applies to the resource manager when handling the hardware components.
+   */
+  std::pair<bool, T> trigger_async_update(
+    const rclcpp::Time & time, const rclcpp::Duration & period)
+  {
+    initialize_async_update_thread();
+    std::unique_lock<std::mutex> lock(async_mtx_, std::try_to_lock);
+    bool trigger_status = false;
+    if (!trigger_in_progress_ && lock.owns_lock())
+    {
+      trigger_in_progress_ = true;
+      current_update_time_ = time;
+      current_update_period_ = period;
+      lock.unlock();
+      async_update_condition_.notify_one();
+      trigger_status = true;
+    }
+    const T return_value = async_update_return_;
```

Comment:
Do you think `std::future` would be more idiomatic here, instead of a condition variable? As I see we're doing some calculations in the async thread, and returning it when it's ready. I think we'd also have one less flag to keep track of.

Reply:
Yeah, I thought so, but I rather prefer that the conditional variable is not awake after a certain duration, because that's exactly what we want to avoid with proper triggering